chemprop/uncertainty/calibrator.py

```diff
@@ -1,13 +1,15 @@
 from abc import abstractmethod
 
+from scipy.optimize import fmin
+import torch
 from torch import Tensor
 
 from chemprop.utils.registry import ClassRegistry
 
 
 class UncertaintyCalibrator:
```

Comment:
broader question on these, can we make these `nn.Module`s?

Reply:
Since it is not a neural network, I would prefer the current way. The only benefit of inheriting from `nnpModule` might be that we can use `nn.Parameter` to register calibration parameters.

Reply:
That's an understatement of the `nn.Module` class. You also device and dtype movement, which are important to prevent `RuntimeError`s.

Reply:
You are right! Since the input of `calibrate()` and `apply_calibration()` is already `torch.Tensor`s, I would expect the dtype and device to be set before passing those values into these functions. Maybe I missed something here. Do you think there is anything we should rely on `nn.Module` to handle during the calibration?

Reply:
The current code needs requires that both `self.a` and `self.b` be moved in two separate lines if the device/dtype needs to be changed at a later point. Seems like it would be cleaner to just call `calibrator.to(...)` rather than:
```python
calibrator.a = calibrator.a.to(...)
calibrator.b = calibrator.b.to(...)
```
for which that snippet only works if you know the parameters which need to be moved. If you have a `ZScalingCalibrator`, then the snippet changes, whereas the `.to(...)` call never will.

Reply:
That is a good point about using `nn.Module`. Inheriting from it and adding all the new parameters using `nn.Parameter` ensures that all attributes are on the same device within `.to()`.

Reply:
@davidegraff I am not sure if it is necessary to support device change in this object. Would it be an issue if we always use the CPU device here?

Reply:
Is there a reason you _don't_ want seamless GPU acceleration?

Reply:
I expect the computation is not very heavy here, so there won't be much benefit from running it on the GPU. Of course, it won't hurt to provide the functionality of moving devices in this object for cases where the GPU can seamlessly accelerate.

Reply:
Which uncertainty calibration methods would benefit from seamless GPU acceleration? I see several of the methods (including Platt) use `fmin` to fit a couple parameters, which I don't think supports GPU. If `a` and `b` have to be registered in an `init` method, and the user does `.to(cuda)` before fitting, would I need to adjust the `.fit` method to move `a` and `b` back to the CPU and then after fitting move the fit parameters back to the GPU?

The `apply_calibration` of Platt could use a GPU, but I don't have a sense for whether transferring the data to the GPU and running the calculation would be faster than just running the calculation on CPU. 

I think the uncertainty calibrators are more like an sklearn `StandardScaler` (which doesn't support GPUs) than they are like a torch Module. They don't have children modules, a forward method, a call method, or a sense of if the object is in training mode. But they do fit a couple parameters to some data to then be applied to another set of data.

Reply:
I'm not sure why there's such opposition to such a small change, but I'm not going continue to debate this point if you feel this strongly so long as there are no device mismatch errors down the road.

Reply:
It's a small change and a step towards GPU acceleration, but my thought is we should limit our scope to only support calibration on CPU. With a limited scope we don't have to worry about moving data/models to a GPU, like in the CLI or in our implementation of the calibrators. Perhaps later someone will want to support calibration on GPU and can do a full implementation. As a quick test to see what the cost/benefits are, I did the following timing test:

I added an init method to register the parameters:
```
class PlattCalibrator(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("a", torch.tensor([]))
        self.register_buffer("b", torch.tensor([]))
```
In `calibrate` (which I will rename `fit` along with other changes later), I initialize a and b as an empty tensor on the correct device instead of an empty list. (This is for the case the user calls `calibrate` multiple times.) 
```
self.a = torch.tensor([]).to(self.a.device)
self.b = torch.tensor([]).to(self.b.device)
```
I also turn the output of fmin to a tensor and concatenate instead of append.
```
a_j = torch.tensor([a_j]).to(self.a.device)
b_j = torch.tensor([b_j]).to(self.b.device)
self.a = torch.cat((self.a, a_j))
self.b = torch.cat((self.b, b_j))
```
The timing code is 
```
cal  = PlattCalibrator()
preds = torch.rand(100000, 8)
targets = torch.randint_like(preds, 0, 2)
mask = torch.isfinite(preds)
```
```
%%timeit
cal.calibrate(preds, targets, mask)
```
```
%%timeit
cal.apply_calibration(preds)
```
```
cal.to("cuda")
preds = preds.to("cuda")
targets = targets.to("cuda")
mask = mask.to("cuda")
```
```
%%timeit
cal.calibrate(preds, targets, mask)
```
```
%%timeit
cal.apply_calibration(preds)
```
With these results
| | CPU | GPU |
|---|---|---|
| fit | 269 ms | 87 ms |
| apply | 500 μs | 64 μs |

So using the GPU is faster, by about 0.2 seconds for a decently large dataset (100000 x 8), which is actually more than I was expecting. Though, as the predictions and targets start on the CPU anyways, my preference is to ignore that GPUs exist when implementing the calibrators to make the implementation easier.